### DISCUSSION TITLE
NoSql Database support

### DISCUSSION THREAD

**Discussion by aashi0001** [jeremyevans/rodauth#285] [General]
Is it possible to integrate Rodauth with an app which is completely based on nosql DB, in my case it’s Dynamodb but if let’s say we consider Dynamodb or MongoDB, Can we introduce an adapter for nosql databases or a way in which Roadauth can directly be used with an app which uses only NoSql Database.

**Comment by jeremyevans** [comment]
There's no way to have a NoSQL adapter. Rodauth is based on Sequel, an SQL database library, and the implementation uses SQL.

You can use alternative databases using Rodauth's configuration methods to override all code that access the database, and have it use a NoSQL database instead.  Depending on how much of Rodauth you are using, that could range from fairly simple (login with password only) to very challenging.

**Reply by aashi0001** [reply]
But wouldn’t it be nice that for future if we could put it in roadmap to support NoSql databases, as Devise currently does support MongoDB, I can also help in this.

what do you think ?

**Reply by jeremyevans** [reply]
Not on the roadmap, never going to be on the roadmap. :)

**Reply by HoneyryderChuck** [reply]
Adding to Jeremy's comment, not only is rodauth using sequel, it's also not using its ORM Model interface, but the dataset API, which is a quasi SQL query builder, so it'd be hard to accomplish even if there was a sequel plugin for mongo. 

Jeremy's suggestion is probably the most reasonable way to implement it. However, even if you managed to rewrite all sql based hooks and made rodauth work with mongo, it'd still be incompatible with most of rodauth external plugin ecosystem (I maintain 2 gems based on rodauth, and I'd argue it'd be quite hard to achieve compatibility in rodauth-oauth). 

Devise has the exact same problem, perhaps even more exacerbated, as the fraction of its external plugin ecosystem that still works, mostly targets vanilla activerecord features, and is largely incompatible with the rest (I've used devise-sequel in the past).
